### ISSUE TITLE
Diffrax has lower than expected accuracy vs scipy's solve_ivp

### ISSUE THREAD

**Issue by TheCHIRIk** [patrick-kidger/diffrax#634]
Thank you for building and maintaining this crucial package!

I am rewriting a plain Python simulator into Jax for RL training, and I need to declare an accuracy tolerance for my version vs the original. The original uses scipy's `solve_ivp`, and I am using Diffrax's `diffeqsolve` as the Jax counterpart. However, I do not get the accuracy tolerances I would expect, such as those declared by `rtol` or `atol`. 

Specifically, let's say `rtol=1e-7`. Let's call the true value of an ODE `x`, while the value returned by diffrax is `x_d`, and scipy's is `x_s`. Assume both of them ran with the required `rtol` to get their values. Since they are both meant to be within `rtol` of `x`, their difference should be within `abs(x_d - x_s) <= rtol * 2 < rtol * 10`. Thus, I would expect that a tolerance one order of magnitude higher than `rtol` will always be met. However, as you can see by failing test cases `(1e-7, 1e-9, 6)` and `(1e-9, 1e-11, 6)` below, this is not always the case.

See this toy test (inspired by [this answer](https://github.com/patrick-kidger/diffrax/issues/82#issuecomment-1080324638)) for a quick showcase:

```python
# %% Imports
import diffrax
import jax
import jax.numpy as jnp
import numpy as np
from scipy.integrate import solve_ivp
from numpy.testing import assert_array_almost_equal
import pytest
@pytest.mark.parametrize(("rtol", "atol", "check_tol"),
                         # These succeed
                         [(1e-3, 1e-5, 3),
                          (1e-6, 1e-9, 5),
                          (1e-7, 1e-9, 5),
                          (1e-9, 1e-11, 5),
                         # The ones below fail
                          (1e-3, 1e-5, 4),
                          (1e-6, 1e-9, 6),
                          (1e-7, 1e-9, 6),
                          (1e-7, 1e-9, 7),
                          (1e-9, 1e-11, 6),
                          ])
def test_random(rtol, atol, check_tol):
    N = 200
    h = 1 / (N + 1)
    y0 = np.ones(N, dtype=np.float32)
    y0[0] = y0[-1] = 0.0
    tspan = (0, 0.01)

    def func(t, y, dy, *, use_jax=False):
        if use_jax:
            idx = jnp.arange(1, N - 1)
            dy = dy.at[idx].set((y[idx - 1] - 2 * y[idx] + y[idx + 1]) / h ** 2)
        else:
            idx = np.arange(1, N - 1)
            dy[idx] = (y[idx - 1] - 2 * y[idx] + y[idx + 1]) / h ** 2
        return dy

    # %% Scipy's solve_ivp
    y0 = np.array(y0)
    dy = np.zeros(N)
    odefunc = lambda t, y: func(t, y, dy)
    solve_scipy = lambda y0: solve_ivp(odefunc, tspan, y0, rtol=rtol, atol=atol)
    s_scipy = solve_scipy(y0)['y'][:, -1]

    # %% Diffrax's diffeqsolve
    jax.config.update("jax_enable_x64", False)
    dy = jnp.zeros(N)
    odefunc = lambda t, y, args: func(t, y, dy, use_jax=True)
    eqs = diffrax.ODETerm(odefunc)
    solver = diffrax.Dopri5()
    stepsize_controller = diffrax.PIDController(rtol=rtol, atol=atol)
    y0 = jnp.array(y0, dtype=dy.dtype)

    @jax.jit
    def solve_diffrax(y0):
        return diffrax.diffeqsolve(eqs, solver, t0=tspan[0], t1=tspan[1], dt0=None, y0=y0,
                                   stepsize_controller=stepsize_controller)
    s_jax = solve_diffrax(y0).ys[-1]

    # %% Compare accuracies
    assert_array_almost_equal(s_jax, s_scipy, decimal=check_tol)
``` 
![Image](https://github.com/user-attachments/assets/a21f6d05-1606-451e-9a5e-f45c1f8713a5)

Could you explain what I am missing with my expectations?

**Comment by patrick-kidger** [comment]
Yup! So `rtol` and `atol` have a slightly unusual meaning in differential equation solvers. They refer to a bound on the _estimated local error_, i.e. the solver's own estimate of how much error it made over a single numerical step.

So in particular these don't provide a guarantee about the quantitative closeness of two solutions.

---

This isn't quite your question, but I suspect it'll come up so I'm going to expand on the above and turn it into a larger teaching point on how you can compare multiple solvers.

From the above, one important detail is that the meaning of these tolerances is actually solver specific. Two different solvers will tend to make very different estimates of their error, and so it isn't actually possible to fix a tolerance and expect to compare two different solvers. (Whether those are e.g. two different diffrax solvers, or e.g. a diffrax vs scipy solver.)

As for how this should actually be done, the answer is a work-precision diagram. Compute a 'true' solution using an expensive solve with very small steps. Then for each solver under consideration, do a sweep over varying tolerances: for each tolerance, measure both (a) the amount of work it took, e.g. wallclock time, and (b) the error to the 'true' solution. You end up with multiple lines on a 2D plot, one line for each solver, with each line corresponding to varying tolerances.

**Comment by patrick-kidger** [comment]
Any solve you like, whether from Diffrax/scipy/etc, with really tiny steps so that it's trusted to be pretty close to the true value.

**Comment by TheCHIRIk** [comment]
Thank you, this is helpful! Feel free to close this as answered.

**Comment by patrick-kidger** [comment]
Will do! Glad I could help :)
